<div style="border:solid Chocolate 2px; padding: 40px">

Привет, меня зовут Александр Куимов. Я буду ревьюером твоего проекта. Ты можешь обращаться ко мне на "ты"😏 Надеюсь, тебя также не смутит, если я буду обращаться к тебе на "ты", но если это неудобно, обязательно скажи об этом!

**Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии**, они будут особенно полезны для нашей работы в случае повторной проверки проекта. 

Прошу реагировать на все мои комментарии, по которым ты вносишь правки в проект, своими комментариями по шаблону, показанному чуть ниже. На каждое мое замечание нужно отвечать комментарием студента в маркдаун ячейке **ниже моей. Это нужно, чтобы не создавалась путаница**.😉

Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным.
</div>

<br/>
 
<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>
    
    
<br/>    
<div class="alert alert-info">
<h2> Комментарий студента: <a class="tocSkip"> </h2>

<b>👋:</b> В такой цветовой ячейке - образце я прошу тебя оставлять твои комментарии. **Обязательно добавь заголовок "Комментарий студента" к каждой твоей ячейке**. Если исправляешь проект во второй раз и далее, не забывай пожалуйста указывать номер итерации, например, "Комментарий студента v.2". У меня к этому моменту уже появится нумерация в моих комментариях.
</div> 

<br/>    
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самому найти и исправить ее. На реальной работе твой руководитель будет поступать также, и я пытаюсь подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку!🤓
    
 

# Проект спринта 7

- Автор: Елагина Ксения
- Дата: 31.03.2025

### Цели и задачи проекта
Необходимо изучить объёмы продаж игр разных жанров и региональные предпочтения игроков. Данный анализ поможет заказчику написать статью о развитии индустрии игр в начале XXI века и тем самым привлечь новую аудиторию к игре «Секреты Темнолесья».

Для этого необходимо:
- Изучить развитие игровой индустрии с 2000 по 2013 год;
- Сделать обзор игровых платформ;
- Изучить объёмы продаж игр разных жанров;
- Изучить региональные предпочтения игроков.

### Описание данных
В проекте будут использованы данные датасета `new_games.csv` с таким описанием:
- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта
1. Загрузка и знакомство с данными
2. Проверка ошибок и предобработка данных
3. Фильтрация данных
4. Категоризация данных
5. Итоговый вывод

---

## 1. Загрузка данных и знакомство с ними

In [1]:
# Импортируем библиотеку pandas:
import pandas as pd

In [2]:
# Выгрузим данные из датасета new_games.csv в датафрейм new_games:
new_games = pd.read_csv('https://code.s3.yandex.net//datasets/new_games.csv')

In [3]:
# Выведем информацию о датафрейме:
new_games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
# Выведем первые строки датафрейма на экран:
new_games.head()

Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

Датасет содержит 11 столбцов и 16956 строк, в которых представлена информация о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр.

Изучим типы данных и их корректность:
- **Числовые значения с плавающей точкой (float64):** 4 столбца представлены типом `float64`:
    - `Year of Release` содержит данные года выпуска игры (для данных подойдет тип `int16`);
    - `Critic Score` содержит данные оценки критиков (для данных подойдет целочисленный тип `int8`, оценки от 0 до 100);
    - `NA sales` и `Other sales` можно изменить на `float32`, чтобы использовать меньший объем памяти (столбцы хранят данные о продажах в Северной Европе и в других странах в млн шт.).


- **Строковые данные (object):** остальные 7 столбцов содержат тип `object`:
    - `EU sales`, `JP sales`, `User Score` содержат данные о продаже копий в Европе и Японии (в млн шт.), оценку пользователей от 0 до 5 (столбцам подойдет тип данных с плавающей точкой - `float32`, в них содержатся численные значения, которые подойдут данным, выбираем минимальное значение бит);
    - `Rating` содержит данные о рейтинге организации ESRB (подойдет категориальный тип `category`, столбец состоит из категорий игр для разных возрастов);
    - `Name`, `Platform`, `Genre` - оставшимся столбцам подходит тип данных `object`, потому что данные состоят из строковой информации (название игры, название платформы, жанр игры).

Названия столбцов лучше привести к единому виду - начинать название с маленькой буквы и заменить пробелы на нижнее подчеркивание ( _ ).

In [5]:
# Посчитаем количество пропусков в столбцах:
new_games.isna().sum()

Name                  2
Platform              0
Year of Release     275
Genre                 2
NA sales              0
EU sales              0
JP sales              0
Other sales           0
Critic Score       8714
User Score         6804
Rating             6871
dtype: int64

Большинство пропусков встречаются в оценках критиков и пользователей, рейтинге ERSB.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 
    
С данными познакомились, проведен первичный аналитический осмотр!) Молодец, что детально изучаешь каждый столбец и указываешь основные ошибки
    
Полученная информация поможет нам в дальнейшем с предобработкой данных
</div>

---

## Проверка ошибок в данных и их предобработка

### Названия столбцов датафрейма

In [6]:
# Выведем названия всех столбцов датафрейма:
new_games.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [7]:
# Приведем все столбцы к единому стилю написания:
new_columns_name = ['name', 'platform', 'year_of_release', 'genre', 'north_america_sales', 'europe_sales',
       'japan_sales', 'other_sales', 'critic_score', 'user_score', 'ersb_rating']
new_games.columns = new_columns_name
new_games.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'north_america_sales',
       'europe_sales', 'japan_sales', 'other_sales', 'critic_score',
       'user_score', 'ersb_rating'],
      dtype='object')

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 
    
Отлично, названия столбцов приведены к стилю snake_case, это делает датафрейм более удобным для работы и чтения кода 
    
В целом тут можно использовать цепочку методов, добавлю другие альтернативные способы переименования названия столбцов:      
        
- чтобы переименовать колонки, можно также использовать метод `rename()`



- также интересным является подход с использованием list comprehensions:
        
`df.columns = [x.lower().replace(' ', '_') for x in df.columns.values]`
    
    
Данный подход использует list comprehensions, что может быть быстрее на больших датасетах благодаря оптимизации Python для list comprehensions. Этот метод также предоставляет гибкость в добавлении дополнительных операций или условий при необходимости. Однако, он может быть менее очевидным для тех, кто только начинает работать с Python. Этот подход "интересен" тем, что он показывает, как одна и та же задача может быть решена разными способами, и некоторые из них могут быть более эффективными или гибкими в зависимости от конкретного контекста   
    
            

### Наличие пропусков в данных

In [8]:
# Посчитаем количество пропусков в столбцах:
new_games.isna().sum()

name                      2
platform                  0
year_of_release         275
genre                     2
north_america_sales       0
europe_sales              0
japan_sales               0
other_sales               0
critic_score           8714
user_score             6804
ersb_rating            6871
dtype: int64

In [9]:
# Посчитаем процент пропусков в столбцах:
new_games.isna().mean()*100

name                    0.011795
platform                0.000000
year_of_release         1.621845
genre                   0.011795
north_america_sales     0.000000
europe_sales            0.000000
japan_sales             0.000000
other_sales             0.000000
critic_score           51.391838
user_score             40.127389
ersb_rating            40.522529
dtype: float64

В данных наблюдаются пропуски в следующих столбцах:
- `name`: в 2 строках (0.01% данных) отсутствуют названия игр. Поскольку это незначительная доля данных, такие строки можно удалить, т.к. они не повлияют на анализ;
- `genre`: в 2 строках (0.01% данных) отсутствует информация о названии жанра. Как и в предыдущем случае, эти строки можно удалить;
- `year_of_release`: в 275 строках (1.6% данных) отсутствует год выпуска игры. Необходимо будет проанализировать игры с 2000 по 2013 годы, поэтому можно заменить пропуски на значение-индикатор; 
- `user_score`: в 6804 строках (40.1% данных) отсутствует информация об оценке пользователей. Вероятно, ни в одном из источников не было данных об оценке игр. Пропуски в этом параметре могут негативно повлиять на анализ популярности игр, заменим их на значение-индикатор;
- `ersb_rating `: в 6871 строках (40.5% данных) отсутствует возрастной рейтинг ESRB. Анализ игр несильно зависит от данного параметра, поэтому пропуски можно оставить;
- `critic_score`: в 8714 строках (51.4% данных) отсутствует информация об оценке критиков. Пропуски составляют более, чем половину датасета, что может затруднить анализ самых популярных игр по оценкам критиков, заменим их на значение-индикатор.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 

Проверить колличество пропусков в % можно еще вот так: 
    
    
    pd.DataFrame(round(df.isna().mean()*100,)).style.background_gradient('coolwarm')
    
- Можно также использовать f-строку для вывода процентов и округления данных :.2%   https://python-scripts.com/f-strings

In [10]:
# Удалим строки с пропусками в name и genre:
new_games = new_games.dropna(subset = ['name', 'genre'])

In [11]:
# Заменим пропуски в year_of_release, user_score и critic_score на значение (-1):
new_games['year_of_release'] = new_games['year_of_release'].fillna(-1)
new_games['user_score'] = new_games['user_score'].fillna(-1)
new_games['critic_score'] = new_games['critic_score'].fillna(-1)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b>  
    
Пропуски были изучены и проведена их обработка. молодец, что используешь индикаторы при заполнении пропусков)    

### 2.2. Типы данных

In [12]:
# Выведем информацию о типе данных всех столбцов датафрейма:
new_games.dtypes

name                    object
platform                object
year_of_release        float64
genre                   object
north_america_sales    float64
europe_sales            object
japan_sales             object
other_sales            float64
critic_score           float64
user_score              object
ersb_rating             object
dtype: object

При знакомстве с данными было установлено, что большинство столбцов имеют неверный тип данных. Поэтому необходимо их преобразовать для дальнейшего анализа.

In [13]:
# Приведем тип float64 к int в столбце year_of_release с понижением разрядности:
new_games['year_of_release'] = pd.to_numeric(new_games['year_of_release'], downcast='integer')

In [14]:
# Приведем тип float64 к int в столбце critic_score с понижением разрядности:
new_games['critic_score'] = pd.to_numeric(new_games['critic_score'], downcast='integer')

In [15]:
# Преобразуем типы данных north_america_sales, other_sales с понижением разрядности:
for column in ['north_america_sales', 'other_sales']:
    new_games[column] = pd.to_numeric(new_games[column], downcast = 'float')

Выведем уникальные значения столбцов `europe_sales`, `japan_sales`, `user_score`, чтобы понять, почему данные имеют тип `object`.

In [16]:
new_games['europe_sales'].unique()

array(['28.96', '3.58', '12.76', '10.93', '8.89', '2.26', '9.14', '9.18',
       '6.94', '0.63', '10.95', '7.47', '6.18', '8.03', '4.89', '8.49',
       '9.09', '0.4', '3.75', '9.2', '4.46', '2.71', '3.44', '5.14',
       '5.49', '3.9', '5.35', '3.17', '5.09', '4.24', '5.04', '5.86',
       '3.68', '4.19', '5.73', '3.59', '4.51', '2.55', '4.02', '4.37',
       '6.31', '3.45', '2.81', '2.85', '3.49', '0.01', '3.35', '2.04',
       '3.07', '3.87', '3.0', '4.82', '3.64', '2.15', '3.69', '2.65',
       '2.56', '3.11', '3.14', '1.94', '1.95', '2.47', '2.28', '3.42',
       '3.63', '2.36', '1.71', '1.85', '2.79', '1.24', '6.12', '1.53',
       '3.47', '2.24', '5.01', '2.01', '1.72', '2.07', '6.42', '3.86',
       '0.45', '3.48', '1.89', '5.75', '2.17', '1.37', '2.35', '1.18',
       '2.11', '1.88', '2.83', '2.99', '2.89', '3.27', '2.22', '2.14',
       '1.45', '1.75', '1.04', '1.77', '3.02', '2.75', '2.16', '1.9',
       '2.59', '2.2', '4.3', '0.93', '2.53', '2.52', '1.79', '1.3', '2.6',
   

In [17]:
new_games['japan_sales'].unique()

array(['3.77', '6.81', '3.79', '3.28', '10.22', '4.22', '6.5', '2.93',
       '4.7', '0.28', '1.93', '4.13', '7.2', '3.6', '0.24', '2.53',
       '0.98', '0.41', '3.54', '4.16', '6.04', '4.18', '3.84', '0.06',
       '0.47', '5.38', '5.32', '5.65', '1.87', '0.13', '3.12', '0.36',
       '0.11', '4.35', '0.65', '0.07', '0.08', '0.49', '0.3', '2.66',
       '2.69', '0.48', '0.38', '5.33', '1.91', '3.96', '3.1', '1.1',
       '1.2', '0.14', '2.54', '2.14', '0.81', '2.12', '0.44', '3.15',
       '1.25', '0.04', '0.0', '2.47', '2.23', '1.69', '0.01', '3.0',
       '0.02', '4.39', '1.98', '0.1', '3.81', '0.05', '2.49', '1.58',
       '3.14', '2.73', '0.66', '0.22', '3.63', '1.45', '1.31', '2.43',
       '0.7', '0.35', '1.4', '0.6', '2.26', '1.42', '1.28', '1.39',
       '0.87', '0.17', '0.94', '0.19', '0.21', '1.6', '0.16', '1.03',
       '0.25', '2.06', '1.49', '1.29', '0.09', '2.87', '0.03', '0.78',
       '0.83', '2.33', '2.02', '1.36', '1.81', '1.97', '0.91', '0.99',
       '0.95', '2.0'

In [18]:
new_games['user_score'].unique()

array(['8', -1, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

Вероятнее всего в открытых источниках, из которых извлекали данные, были допущены ошибки. Поэтому среди численных значений попались строковые. Преобразуем тип данных `object` к `float`:

In [19]:
# Преобразуем тип данных object к float в столбцах europe_sales, japan_sales, user_score с понижением разрядности:
for column in ['europe_sales', 'japan_sales', 'user_score']:
    new_games[column] = pd.to_numeric(new_games[column], errors = 'coerce', downcast='float')

In [20]:
# Проверим результаты преобразований:
new_games[['year_of_release','critic_score','north_america_sales','other_sales',
 'europe_sales','japan_sales', 'user_score']].dtypes

year_of_release          int16
critic_score              int8
north_america_sales    float32
other_sales            float32
europe_sales           float32
japan_sales            float32
user_score             float32
dtype: object

Для оптимизации работы с данными в датафрейме были сделаны такие изменения типов данных:

- `year_of_release`: тип данных изменён с `float64` на `int16`;
- `critic_score`: тип данных изменён с `float64` на `int8`;
- `north_america_sales`, `other_sales`: тип данных изменён с `float64` на `float32`;
- `europe_sales`, `japan_sales`, `user_score`: тип данных изменён с `object` на `float32`.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 
    
Использование pd.to_numeric() с параметром errors='coerce' для обработки некорректных значений — это правильное решение, которое позволяет избежать ошибок при преобразовании 
    
<div class="alert alert-warning">

    
<b>Некоторые замечания и рекомендации💡:</b> 
    
Перед проведением обработки типов данных желательно посмотреть на уникальные значения в столбцах eu_sales, jp_sales и user_score, далее найти причины некорректных типов. Затем уже можно принимать решение об обработке    

<br/>    
<div class="alert alert-info">
<h2> Комментарий студентки v.2: <a class="tocSkip"> </h2>

<b>👋:</b> Я приняла к сведению рекомендации и добавила информацию об уникальных значениях [16]-[18].
</div> 

<div class="alert alert-success">
<h2> Комментарий ревьюера 2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>

### 2.4. Явные и неявные дубликаты в данных


Изучим уникальные значения в столбцах `platform`, `genre`, `user_score` и `ersb_rating`:

In [21]:
new_games['platform'].unique()

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [22]:
new_games['genre'].unique()

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

In [23]:
new_games['ersb_rating'].unique()

array(['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

Среди названий жанров присутствуют неявные дубликаты. Необходимо будет привести все данные к единому виду, чтобы избежать повторов. В других столбцах повторы отсутствуют.

In [24]:
# Приведем значения в столбце genre к нижнему регистру:
new_games['genre'] = new_games['genre'].str.lower()
new_games['genre']

0              sports
1            platform
2              racing
3              sports
4        role-playing
             ...     
16951          action
16952          sports
16953       adventure
16954        platform
16955      simulation
Name: genre, Length: 16954, dtype: object

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера  <a class="tocSkip"></h2>

    
<b>На доработку🤔:</b> 

Приводить признаки к нижнему/верхнему регистру нужно с перезаписыванием, например:

    df['...'] = df['...'].str.lower()
  

<br/>    
<div class="alert alert-info">
<h2> Комментарий студентки v.2: <a class="tocSkip"> </h2>

<b>👋:</b> Исправила [24].
</div> 

<div class="alert alert-success">
<h2> Комментарий ревьюера 2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>

In [25]:
# Проверим наличие явных дубликатов:
new_games.duplicated().sum()

np.int64(241)

In [26]:
# Удалим явные дубликаты:
new_games.drop_duplicates(inplace=True)

In [27]:
# Проверим результат преобразования:
new_games.duplicated().sum()

np.int64(0)

Неявные дубликаты присутствовали только в столбце `genre`, данные которого были приведены к нижнему регистру.
Явные дубликаты были найдены в 182 строках, после чего их удалили.

**К данному этапу была проведена полная предобработка данных:**
- все столбцы привели к единому стилю;
- проверили данные на наличие пропусков и провели преобразования для исключения ошибок при дальнейшем анализе;
- преобразовали столбцы к правильным типам данных;
- нашли явные и неявные дубликаты и избавились от них.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 

Методы  `.duplicated()` и `.drop_duplicates()` использованы корректно)
    
можно еще подсчитать процент/долю отфильтрованных данных   

---

## 3. Фильтрация данных

Для анализа необходимо провести фильтрацию с 2000 по 2013 год включительно.

In [28]:
# Создадим срез данных с 2000 по 2013 год:
df_actual = new_games.loc[(new_games['year_of_release'] >= 2000)&(new_games['year_of_release'] <= 2013)].copy()

In [29]:
# Проверим результаты фильтрации:
df_actual['year_of_release'].unique()

array([2006, 2008, 2009, 2005, 2007, 2010, 2013, 2004, 2002, 2001, 2011,
       2012, 2003, 2000], dtype=int16)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 

Фильтрация данных проведена корректно)  Можно еще посмотреть, сколько значений было удалено

---

## 4. Категоризация данных
    
Проведем категоризацию данных:
1. Разделим все игры по оценкам пользователей (`user_score`) и выделим такие категории: `высокая оценка` (от 8 до 10 включительно), `средняя оценка` (от 3 до 8, не включая правую границу интервала) и `низкая оценка` (от 0 до 3, не включая правую границу интервала).

    Все пропуски, замененные на значения-индикаторы, отнесем к группе `нет оценки`.

In [30]:
# Разделим игры на категории по user_score:
df_actual['user_score_category'] = pd.cut(df_actual['user_score'], bins = [-2, 0, 2, 7, 10], labels = ["нет оценки", "низкая оценка", "средняя оценка", "высокая оценка"])


2. Разделим все игры по оценкам критиков (`critic_score`) и выделите такие категории: `высокая оценка` (от 80 до 100 включительно), `средняя оценка` (от 30 до 80, не включая правую границу интервала) и `низкая оценка` (от 0 до 30, не включая правую границу интервала).

   Все пропуски, замененные на значения-индикаторы, отнесем к группе `нет оценки`.

In [31]:
# Разделим игры на категории по critic_score:
df_actual['critic_score_category'] = pd.cut(df_actual['critic_score'], bins = [-2, 0, 29, 79, 100], labels = ["нет оценки", "низкая оценка", "средняя оценка", "высокая оценка"])

- Составим пользовательскую функцию, которая поможет сгруппировать данные по категориям и провести сравнение по количеству и долям игр:

In [32]:
# Составим функцию для группировки и категоризации данных:
def score_category(df, category):
    grouped_category = df.groupby([category], observed=False)['name'].count()
    if 'нет оценки' in grouped_category:
        grouped_category = grouped_category.drop('нет оценки')

    top_category = grouped_category.sort_values(ascending=False)
    top_part_category = top_category/top_category.sum()*100
    return top_category, top_part_category 

In [33]:
# Применим функцию к user_score и сравним по количеству игр:
top_user_score_category, top_part_of_user_score_category = score_category(df_actual, 'user_score_category')
top_user_score_category

user_score_category
высокая оценка    4085
средняя оценка    2351
низкая оценка       46
Name: name, dtype: int64

In [34]:
# Сравним по процентам:
top_part_of_user_score_category

user_score_category
высокая оценка    63.020673
средняя оценка    36.269670
низкая оценка      0.709658
Name: name, dtype: float64

In [35]:
# Применим функцию к critic_score и сравним по количеству игр:
top_critic_score_category, top_part_of_critic_score_category = score_category(df_actual, 'critic_score_category')
top_critic_score_category

critic_score_category
средняя оценка    5422
высокая оценка    1692
низкая оценка       55
Name: name, dtype: int64

In [36]:
# Сравним по процентам:
top_part_of_critic_score_category

critic_score_category
средняя оценка    75.631190
высокая оценка    23.601618
низкая оценка      0.767192
Name: name, dtype: float64

Большинство игр, выпущенных с 2000 по 2013 г. получили высокую оценку от пользователей - 62% от всего количества.

Критики же были иного мнения и отнесли большинство игр - целых 83% - к категории со средней оценкой. Высокие оценки по мнению критиков заслужили лишь 26% от всех игр. Это количество почти в 3 раза меньше, нежели оценка пользователей.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера  <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
        
Категоризация выполнена верно, как вариант можно было написать функции для категоризации с целью уменьшения дублирования кода и его унификации, далее использовать функции с помощью метода .apply() 

<br/>    
<div class="alert alert-info">
<h2> Комментарий студентки v.2: <a class="tocSkip"> </h2>

<b>👋:</b> Исправила [32]-[36].
</div> 

<div class="alert alert-success">
<h2> Комментарий ревьюера 2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b>

- Выделим топ-7 платформ по количеству игр, выпущенных за весь актуальный период:

In [37]:
# Отсортируем платформы по количеству игр:
top_platform = df_actual['platform'].value_counts()

In [38]:
# Выведем топ-7 платформ по количеству игр, выпущенных с 2000 по 2013 г:
top_platform.head(7)

platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: count, dtype: int64

In [39]:
# Отсортируем платформы в долях:
top_platform_part = df_actual['platform'].value_counts(normalize=True)

In [40]:
# Выведем топ-7 платформ в процентах:
top_platform_part.head(7)*100

platform
PS2     16.641890
DS      16.587122
Wii      9.975745
PSP      9.232454
X360     8.770832
PS3      8.504812
GBA      6.345356
Name: proportion, dtype: float64

Было выявлено, что чаще всего с 2000 по 2013 гг. игры выпускали на PS2 и DS - свыше, чем 16% от всего количества платформ. Игры на этих консолях были выпущены почти в 2 раза больше, чем на платформах, которые занимают 3-7 места. Так Wii И PSP составляют чуть больше 9%.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 

Отлично, что выведен ТОП-7 платформ по количеству игр)

---

## 5. Итоговый вывод

В данном проекте были изучены данные `new_games.csv` о продажах игр разных жанрах и платформ, а также пользовательские и экспертные оценки игр. Датасет состоит из 11 столбцов и 16956 строк. При первичном знакомстве с данными и их предобработке получили такие результаты:
- Все названия столбцов были приведены к единому виду - первая буква стала начинаться со строчной, а пробелы заменили нижним подчеркиванием;
- В 6 столбцах (`name`, `genre`, `year_of_release`, `user_score`, `ersb_rating `, `critic_score`) были обнаружены пропущенные значения. Максимальное значение пропущенных данных оказалось в столбце `critic_score` - 51%.
- Для оптимизации работы с данными в датафрейме были произведены следующие изменения типов данных:
    - `year_of_release`: тип данных изменён с `float64` на `int16`;
    - `critic_score`: тип данных изменён с `float64` на `int8`;
    - `north_america_sales`, `other_sales`: тип данных изменён с `float64` на `float32`;
    - `europe_sales`, `japan_sales`, `user_score`: тип данных изменён с `object` на `float32`.
- Для исключения неявных дубликатов названия жанров привели к нижнему регистру;
- Явные дубликаты были удалены в 182 строках;
- С помощью фильтрации был выделен необходимый для анализа срез данных - с 2000 по 2013 гг;
- Сравнение игр по категориям оценки пользователей и критиков выявило, что пользователи поставили высокую оценку 62% игр, когда как критики - только 26%. Критики большей части игр в 83% поставили средний балл;
- Составленный топ-7 платформ по выпуску игр показал, что чаще всего игры выпускали на PS2 и DS - свыше 16% от всего количества платформ.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:  </b> 

Молодец, что не забываешь написать итоговый вывод по проекту. И отлично, что  пишешь выводы по предобработке — это хорошая практика, используй пожалуйста ее в следующих проектах, когда будешь готовить данные 

</div>   
    
    
<div style="border:solid Chocolate 2px; padding: 40px">


<h2> Итоговый комментарий ревьюера (общий вывод по проекту)  <a class="tocSkip"> </h2>    
    
  
У меня сложилось хорошее впечатление о проекте. Молодец! Осмысленная аналитика и дельная работа - многое удалось как надо)
    
Отмечу отдельные положительные моменты проекта🙂:
    
- Была проделана значительная работа по обработке и анализу данных! 
- Твой вывод чётко отражает проделанные шаги, что важно для демонстрации результатов исследования.   
    
Есть несколько моментов всего, на которые стоит ещё раз взглянуть, я указал их в моих комментариях по ходу проекта.
    
Предлагаю тебе доработать проект по моим комментариям, чтобы довести его до совершенства.   
    

 
</div>


<div style="border:solid Chocolate 2px; padding: 40px">


<h2>  Итоговый комментарий ревьюера 2 (итоговый вывод по проекту) <a class="tocSkip"> </h2>    
    
      
  
Теперь почти идеально, молодец! Принимаю твой проект)
    
Могу порекомендовать тебе отличную книгу про Python:
    
https://monster-book.com/avtomatizaciya-zadach-s-python    
    
По визуализации могу посоветовать следующие материалы:
    
1. https://github.com/esokolov/ml-course-hse/blob/master/2022-fall/seminars/sem02-charts.ipynb
    
2. https://github.com/Palladain/Deep_Python/blob/main/Lectures/Lecture_3/Lecture_3_Python.ipynb    
    
Советую тебе обратить внимание на библиотеки `sweetviz` и `ydata_profiling`, которые помогут в проведении более тщательного EDA анализа. Данные библиотеки хороши для максимизации комфорта презентации результатов анализа бизнес-пользователям. 
    
Очень просты в использоовании, на вход кладется датафрейм: `ydata_profiling.ProfileReport(df)`
    
Цель нашего проекта достигнута. Рад был тебе помочь)
    
Удачи тебе в следующем проекте!
    
**Буду очень благодарен, если ты поставишь мне положительную оценку за проект в форме на странице курса, это важно для меня. Достаточно поставить оценку, а комментарий можешь написать по желанию:)**     
    
</div>    
